### DISCUSSION TITLE
parallel_computation.py cannot do parallel computation normally, and repeatition occurs

### DISCUSSION THREAD

**Discussion by HANSIeltsKing** [su2code/SU2#1689] [Q&A] [answered]
Hi, this is an SU2 user who want to use parallel_computation.py to simulate the QuickStart configuration file.
When I access the QuickStart folder and type “parallel_computation.py -f inv_NACA0012.cfg -n 4”, the parallel computation do not work but everything repeat 4 times. The same solution also appear 4 times and the speed become lower.

```
|           0|  1.1562e+00|   -1.026860|    0.201330|    0.082117|
|           1|  1.0859e+00|   -1.169122|    0.263458|    0.037718|
|           1|  1.1016e+00|   -1.169122|    0.263458|    0.037718|
|           1|  1.1172e+00|   -1.169122|    0.263458|    0.037718|
|           1|  1.1562e+00|   -1.169122|    0.263458|    0.037718|
|           2|  1.0833e+00|   -1.236203|    0.301817|    0.029207|
|           2|  1.1042e+00|   -1.236203|    0.301817|    0.029207|
|           2|  1.1354e+00|   -1.236203|    0.301817|    0.029207|
|           2|  1.1458e+00|   -1.236203|    0.301817|    0.029207|
|           3|  1.0742e+00|   -1.327201|    0.320840|    0.024965|
|           3|  1.1055e+00|   -1.327201|    0.320840|    0.024965|
|           3|  1.1172e+00|   -1.327201|    0.320840|    0.024965|
|           3|  1.1172e+00|   -1.327201|    0.320840|    0.024965|
|           4|  1.0906e+00|   -1.419001|    0.329132|    0.023806|
|           4|  1.0844e+00|   -1.419001|    0.329132|    0.023806|
|           4|  1.1125e+00|   -1.419001|    0.329132|    0.023806|
|           4|  1.1187e+00|   -1.419001|    0.329132|    0.023806|
|           5|  1.0859e+00|   -1.513907|    0.332101|    0.023194|
```

However, when I type this "mpirun -np 4 SU2_CFD inv_NACA0012.cfg" the parallel computing can work normally.
I am sure I installed mpich and enabled python wrapper.
How can I solve this problem of those repeatition?

Thank you very much!

**Comment by pcarruscag** [comment]
The parallel_computation script is useless at the moment.
In theory this script will call mpirun, so unless there is another mpirun on your system, which python finds first somehow... it should work.
Try setting SU2_MPI_COMMAND="full path to the mpirun that works" (get the path with `which mpirun`)

**Reply by HANSIeltsKing** [reply]
Thanks for your reply.

After I set SU2_MPI_COMMAND in interface.py by using the path from "which mpirun", the repeatition still occurs. Any code related to python cannot work normally (such as parallel_conputation.py, shape_optimization.py, SU2_CFD.py etc.)

Actually I followed the installation guide step by step, the custom mpi and the python wrapper were also enabled. Could this be any other missing detail in the installation process? The SU2 version I use is 7.3.0.

Thank you very much.

**Comment by thw1021** [comment]
I recently met the same problem. 

Firstly, check your `mpicc` and `mpicxx` path. Try `which mpicc` and `which mpicxx`. The output should be: `/usr/bin/mpicc` and `/usr/bin/mpicxx`. 

If you get something like `/home/username/miniconda3/envs/su2/bin/mpicc` and `/home/username/miniconda3/envs/su2/bin/mpicxx`, then the problem will occur. 

I found the key to solve this problem is to make sure `mpicc` points to `/usr/bin/mpicc` and `mpicxx` points to `/usr/bin/mpicxx`.

Here is the solution I found:

1. uninstall the previously installed `mpi4py`: `conda uninstall mpi4py`
2. reinstall `mpi4py` by `python -m pip install mpi4py --user`

The problem should then be solved. Hope this can help.

**Reply by thw1021** [reply]
Of course, always to make sure that you are using a virtual environment.

**Reply by HANSIeltsKing** [reply]
> 

Thanks for your reply.

I am not sure how to confirm the path is /usr/bin/mpicc?
because the command of "--enable-mpi --with-cc=/usr/bin/mpicc --with-cxx=/usr/bin/mpicxx“ cannot be used in the meson build.

so which command should I type to ensure the meaning of "--enable-mpi --with-cc=/usr/bin/mpicc --with-cxx=/usr/bin/mpicxx"?
otherwise ninja cannot install due to lacking of <mpi.h>

**Reply by pcarruscag** [reply]
export CC=..../mpicc
export CXX=..../mpicxx
Before the first call to ./meson.py

**Reply by HANSIeltsKing** [reply]
> 

but if i do it in the virtual environment, meson cannot build because of the absolute path and relative path mixed.
how can i solve this problem?

**Reply by thw1021** [reply]
This explanation may be better:

make sure `mpicc` and `mpicxx` points to system `mpi` instead of `mpi4py`

I installed `openmpi` by `sudo apt-get install libopenmpi-dev openmpi-bin`. After that, `which mpicc` returns `/usr/bin/mpicc`, and `which mpicxx` returns `/usr/bin/mpicxx`.

**Comment by HANSIeltsKing** [comment]
The current installed version only has SU2_CFD as a parallel format supported by openmpi. But other parts, including SU2_CFD_AD, are still solved in series.

So this is not purely a python problem.

Have you ever tried to successfully parallelize calculations in a solver other than SU2_CFD?

**Comment by thw1021** [comment]
I did a test following [this tutorial](https://su2code.github.io/docs_v7/Quick-Start/).

The Steps:

1. `mpirun -n 2 SU2_CFD inv_NACA0012.cfg > su2_cfd.log 2>&1`, and here is [su2_cfd.log](https://github.com/su2code/SU2/files/9033940/su2_cfd.log)

2. `mpirun -n 2 SU2_CFD_AD inv_NACA0012.cfg > su2_cfd_ad.log 2>&1`, and here is [su2_cfd_ad.log](https://github.com/su2code/SU2/files/9033941/su2_cfd_ad.log)

Things seem fine. Could you have a check, please?

**Comment by thw1021** [comment, marked as answer]
I made a demo to show all I have done when install SU2. Hope it helps.

Please see https://github.com/thw1021/mySU2Docker

**Comment by HANSIeltsKing** [comment]
it works for me
thank you @thw1021

**Reply by thw1021** [reply]
Glad it helps :-)
